File: CDAC/DS_Team_6/PriorityQueue.cpp
#include<iostream>
#include<stdio.h>
using namespace std;
class Data{
    private:
        int data;
        int priority;
    public:
        Data(){
            data = 0;
            priority = 0;
        }

        Data(int data, int priority){
            this->data = data;
            this->priority = priority;
        }

        void setData(int data){
            this->data = data;
        }

        int getData(){
            return data;
        }

        void setPriority(int priority){
            this->priority = priority;
        }

        int getPriority(){
            return priority;
        }
};
typedef Data D;
class Queue{
    private:
        int size;
        D *arr;
        int front;
        int rear;
    public:
        Queue(int size){
            this->size = size;
            arr = new D[size];
            front = rear = -1;
        }

        bool isEmpty(){
            if((front = -1 && rear == -1) || front > rear){
                return true;
            }else{
                return false;
            }
        }

        bool isFull(){
            if(rear == size-1){
                return true;
            }else{
                return false;
            }
        }

        void enqueue(D data){
           if(!isFull()){
               if(front == -1 && rear == -1){
                   front++;
                   rear++;
                   arr[rear] = data;
               }else{
                   int i;
                   for(i = rear;i>=front;i--){
                       if(data.getPriority()>arr[i].getPriority()){
                           arr[i+1]=arr[i];
                       }else{
                           break;
                       }
                   }
                   rear++;
                   arr[i+1] = data;
               }
               cout<<data.getData()<<" enqueued successfully!!!"<<endl;
           }else{
               cout<<"Queue is full!!!"<<endl;
           }
        }

        void dequeue(){
            if(!isEmpty()){
                cout<<"Dequeued element is "<<arr[front].getData()<<endl;
                front++;
            }else{
                cout<<"Queue is empty!!!"<<endl;
            }
        }

        void display(){
            if(!isEmpty()){
                cout<<"Elements in queue are ";
                for(int i=front;i<=rear;i++){
                    cout<<arr[i].getData()<<" ";
                }
                cout<<endl;
            }else{
                cout<<"Queue is empty!!!"<<endl;
            }
        }
};
int main(){
    int size,data,priority,choice;
    cout<<"Enter the size of queue: ";
    cin>>size;
    Data d;
    Queue q(size);
    while(1){
        cout<<"1. Enqueue\n2. Dequeue\n3. Display\n4. Exit"<<endl;
        cout<<"Enter your choice:";
        cin>>choice;
        switch (choice)
        {
            case 1:
                cout<<"Enter the data: ";
                cin>>data;
                cout<<"Enter the priority: ";
                cin>>priority;
                d.setData(data);
                d.setPriority(priority);
                q.enqueue(d);
                break;
            case 2:
                q.dequeue();
                break;
            case 3:
                q.display();
                break;
            case 4:
                exit(0);
                break;
            default:
                cout<<"Enter the correct choice!!!"<<endl;
                break;
        }
    }
}

File: CDAC/edac_data_structure/src/edac_data_structure/PriorityQueue.java
package edac_data_structure;

import java.util.Scanner;

public class PriorityQueue {

	static class Node{
		String data;
		int priority;
		Node(String data,int priority){
			this.data = data;
			this.priority=priority;
		}
	}
	private Node arr[];
	private int size;
	private int front;
	private int rear;
	
	PriorityQueue(){
		arr = new Node[0];
		size=0;
		front=rear=-1;
	}
	
	PriorityQueue(int size){
		this.size=size;
		arr= new Node[size];
		front=rear=-1;
	}
	
	void enqueue(String data,int priority) {
		Node node=new Node(data,priority);
		if(rear == size-1) {
			System.out.println("Queue is full!!!");
		}
		else if(front == -1 && rear == -1) {
			front++;
			rear++;
			arr[rear]=node;
			System.out.println(data+" added successfully!!!");
		}else {
			int i;
			for(i=rear;i>=front;i--) {
				if(priority<arr[i].priority) {
					arr[i+1]=arr[i];
				}else
					break;
			}
			arr[i+1]=node;
			rear++;
			System.out.println(data+" added successfully!!!");
		}
	}
	
	void dequeue() {
		if(front == -1 && rear == -1)
			System.out.println("Queue is empty!!!");
		else if(front>rear) {
			System.out.println("Queue is empty!!!");
		}else {
			System.out.println(arr[front].data+" removed successfully!!!");
			front++;
		}
	}
	
	void print() {
		if(front == -1 && rear == -1)
			System.out.println("Queue is empty!!!");
		else if(front>rear) {
			System.out.println("Queue is empty!!!");
		}else {
			for(int i=front;i<=rear;i++) {
				System.out.print(arr[i].data+" ");
			}
			System.out.println();
		}
		
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter the size of queue:");
		int size = sc.nextInt();
		PriorityQueue q = new PriorityQueue(size);
		int priority,choice=0;
		String data;
		
		while(choice != 4) {
			System.out.println("1. Enqueue\n2. Dequeue\n3. Print\n4. Exit");
			System.out.println("Enter your choice:");
			choice = sc.nextInt();
			switch(choice) {
			case 1:
				System.out.println("Enter the data:");
				data = sc.next();
				System.out.println("Enter the priority:");
				priority = sc.nextInt();
				q.enqueue(data,priority);
				break;
			case 2:
				q.dequeue();
				break;
			case 3:
				q.print();
				break;
			case 4:
				System.exit(0);
				break;
			default:
				System.out.println("Enter valid choice");
			}
		}
		sc.close();
	}
}
